File: com.teaminabox.eclipse.sortit/src/com/teaminabox/eclipse/sortit/SelectionProcessor.scala
package com.teaminabox.eclipse.sortit
import org.eclipse.jface.text.{ ITextSelection, IDocument }
import org.eclipse.ui.{ PlatformUI, IEditorPart }
import org.eclipse.ui.texteditor.ITextEditor

object ProcessSelection {
  def apply(processor: ActionDelegate) {
    val activeEditor = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage().getActiveEditor()
    if (activeEditor.isInstanceOf[ITextEditor]) {
      new SelectionProcessor(processor, activeEditor.asInstanceOf[ITextEditor]).processSelection()
    }
  }

  class SelectionProcessor(processor: ActionDelegate, textEditor: ITextEditor) {

    private val document = textEditor.getDocumentProvider().getDocument(textEditor.getEditorInput())
    private val selection = textEditor.getSelectionProvider().getSelection().asInstanceOf[ITextSelection]
    private var startLineIndex = selection.getStartLine()
    private var endLineIndex = selection.getEndLine()

    if (startLineIndex == endLineIndex) {
      startLineIndex = 0
      endLineIndex = document.getNumberOfLines() - 1
    }

    def processSelection() {
      val processedlines = processor.process(collectLines())
      val startOffset = document.getLineOffset(startLineIndex)
      val replacementText = getReplacementLines(document, startLineIndex, endLineIndex, processedlines)
      document.replace(startOffset, replacementText.length(), replacementText)
      textEditor.getSelectionProvider().setSelection(selection)
    }

    private def getReplacementLines(doc: IDocument, startLineIndex: Int, endLineIndex: Int, lines: List[String]): String = {
      val replacementText = new StringBuffer()
      for (i <- 0 until lines.length) {
        replacementText.append(lines(i))
        if (i < lines.length - 1 || i == lines.length - 1 && doc.getLineDelimiter(endLineIndex) != null) {
          replacementText.append(doc.getLineDelimiter(startLineIndex))
        }
      }
      replacementText.toString()
    }

    private def collectLines(): List[String] = {
      val collectedLines = new scala.collection.mutable.ArrayBuffer[String]()
      for (line <- startLineIndex until endLineIndex + 1) {
        val delimiterLength = if (document.getLineDelimiter(line) == null) 0 else document.getLineDelimiter(line).length()
        val lineText = document.get(document.getLineOffset(line), document.getLineLength(line) - delimiterLength)
        collectedLines += lineText
      }
      collectedLines.toList
    }
  }
}

File: com.teaminabox.eclipse.sortit/src/com/teaminabox/eclipse/sortit/SelectionProcessor.java
package com.teaminabox.eclipse.sortit;

import java.util.ArrayList;

import org.eclipse.jface.text.BadLocationException;
import org.eclipse.jface.text.IDocument;
import org.eclipse.jface.text.ITextSelection;
import org.eclipse.ui.texteditor.ITextEditor;

class SelectionProcessor {

	private ActionDelegate processor;
	private ITextEditor textEditor;
	private IDocument document;
	private ITextSelection selection;
	private int startLineIndex;
	private int endLineIndex;

	public SelectionProcessor(ActionDelegate processor, ITextEditor textEditor) {
		this.processor = processor;
		this.textEditor = textEditor;
		document = textEditor.getDocumentProvider().getDocument(textEditor.getEditorInput());
		selection = (ITextSelection) textEditor.getSelectionProvider().getSelection();
		startLineIndex = selection.getStartLine();
		endLineIndex = selection.getEndLine();

		if (startLineIndex == endLineIndex) {
			startLineIndex = 0;
			endLineIndex = document.getNumberOfLines() - 1;
		}
	}

	public void processSelection() throws BadLocationException {
      ArrayList<String> processedlines = processor.process(collectLines());
      int startOffset = document.getLineOffset(startLineIndex);
      String replacementText = getReplacementLines(document, startLineIndex, endLineIndex, processedlines);
      document.replace(startOffset, replacementText.length(), replacementText);
      textEditor.getSelectionProvider().setSelection(selection);
    }

	private String getReplacementLines(IDocument doc, int startLineIndex, int endLineIndex, ArrayList<String> lines) throws BadLocationException {
      StringBuffer replacementText = new StringBuffer();
      for (int i = 0; i < lines.size(); i++) {		
        replacementText.append(lines.get(i));
        if (i < lines.size() - 1 || i == lines.size() - 1 && doc.getLineDelimiter(endLineIndex) != null) {
          replacementText.append(doc.getLineDelimiter(startLineIndex));
        }
      }
      return replacementText.toString();
    }

	private ArrayList<String> collectLines() throws BadLocationException {
      ArrayList<String> collectedLines = new ArrayList<String>();
      for (int line = startLineIndex; line <= endLineIndex; line++) {
        int delimiterLength;
        if (document.getLineDelimiter(line) == null)
        	delimiterLength = 0;
        else 
        	delimiterLength = document.getLineDelimiter(line).length();
        
        String lineText = document.get(document.getLineOffset(line), document.getLineLength(line) - delimiterLength);
        collectedLines.add(lineText);
      }
      return collectedLines;
    }
  }